gammapy/modeling/models/core.py

```diff
@@ -182,22 +183,43 @@ def __str__(self):
 
     def __add__(self, other):
         if isinstance(other, (Models, list)):
+            dupl = [_ for _ in other if _.name in self.names]
+            if dupl != []:
+                raise (ValueError("Model names must be unique"))
             return Models([*self, *other])
         elif isinstance(other, Model):
+            if other.name in self.names:
+                raise (ValueError("Model names must be unique"))
             return Models([*self, other])
         else:
             raise TypeError(f"Invalid type: {other!r}")
 
-    def __getitem__(self, val):
-        if isinstance(val, int):
-            return self._models[val]
-        elif isinstance(val, str):
+    def __getitem__(self, key):
+        return self._models[self._get_idx(key)]
+
+    def __delitem__(self, key):
+        del self._models[self._get_idx(key)]
+
+    def __setitem__(self, key, model):
+        if model.name in self.names:
```

Comment:
I think we check the mode instance here like `if isinstance(model, (SkyModel, SkyDiffuseCube))` and throw an error else.